scipy/linalg/decomp.py

```diff
@@ -371,126 +415,170 @@ def eigh(a, b=None, lower=True, eigvals_only=False, overwrite_a=False,
-            driver = pfx+'gvd'
-            (gvd,) = get_lapack_funcs((driver,), (a1, b1))
-            v, w, info = gvd(a1, b1, uplo=uplo, itype=type, jobz=_job,
-                             overwrite_a=overwrite_a,
-                             overwrite_b=overwrite_b)
-        # Use '*gv' routine if turbo is off and no eigvals are specified
+        drv_args.update({'lower': lower, 'compute_v': 0 if _job == "N" else 1})
+        w, v, *other_args, info = drv(a=a1, **drv_args, **lwork_args)
+
+    else:  # Generalized problem
+        # 'gvd' doesn't have lwork query
+        if driver == "gvd":
+            drv = get_lapack_funcs(pfx + "gvd", [a1, b1])
+            lwork_args = {}
         else:
-            driver = pfx+'gv'
-            (gv,) = get_lapack_funcs((driver,), (a1, b1))
-            v, w, info = gv(a1, b1, uplo=uplo, itype=type, jobz=_job,
-                            overwrite_a=overwrite_a,
-                            overwrite_b=overwrite_b)
+            drv, drvlw = get_lapack_funcs((pfx + driver, pfx+driver+'_lwork'),
+                                          [a1, b1])
+            # generalized drivers use uplo instead of lower
+            lw = _compute_lwork(drvlw, n, uplo=uplo)
+            lwork_args = {'lwork': lw}
+
+        drv_args.update({'uplo': uplo, 'jobz': _job})
+
+        w, v, *other_args, info = drv(a=a1, b=b1, **drv_args, **lwork_args)
+
+    # m is always the first extra argument
+    w = w[:other_args[0]] if subset else w
+    v = v[:, :other_args[0]] if (subset and not eigvals_only) else v
 
     # Check if we had a  successful exit
     if info == 0:
         if eigvals_only:
             return w
         else:
             return w, v
-    _check_info(info, driver, positive=False)  # triage more specifically
-    if info > 0 and b1 is None:
-        raise LinAlgError("unrecoverable internal error.")
-
-    # The algorithm failed to converge.
-    elif 0 < info <= b1.shape[0]:
-        if eigvals is not None:
-            raise LinAlgError("the eigenvectors %s failed to"
-                              " converge." % nonzero(ifail)-1)
+    else:
+        if info < -1:
+            raise LinAlgError('Illegal value in argument {} of internal {}'
+                              ''.format(-info, drv.typecode + pfx + driver))
+        elif info > n:
+            raise LinAlgError('The leading minor of order {} of B is not '
+                              'positive definite. The factorization of B '
+                              'could not be completed and no eigenvalues '
+                              'or eigenvectors were computed.'.format(info-n))
         else:
-            raise LinAlgError("internal fortran routine failed to converge: "
-                              "%i off-diagonal elements of an "
-                              "intermediate tridiagonal form did not converge"
-                              " to zero." % info)
+            drv_err = {'ev': 'The algorithm failed to converge; {} '
```

Comment:
If the keys were 'v', 'vx', 'vd', and 'vr', could you just:
`msg = drv_err[driver[1:]].format(info)`
(with one special case, apparently)?

Reply:
I wanted to do that but seemed already cryptic so did the laborious way.

Reply:
Ok, it's up to you.